src/server/pfs/server/pathlock.go

```diff
@@ -0,0 +1,156 @@
+package server
+
+import (
+	gopath "path"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/pachyderm/pachyderm/src/client/limit"
+	"github.com/pachyderm/pachyderm/src/client/pkg/errors"
+)
+
+// pathlock is used to prevent two PutFile calls that affect the same path (e.g.
+// Put("/foo") and Delete("/")) from racing. The rule is: before running any
+// PutFile operation B, make sure there is no running PutFile operation A such
+// that B.Path is a prefix of A.Path or vice-versa.
+//
+// Note that pathlock can only block one goro at a time. This works for PutFile
+// (where requests are arriving in a stream, and the whole stream stream can be
+// paused until conflicting paths are finished), but you could not use this to
+// schedule N goroutines and have each independently call start() (i.e. many
+// goros blocking at once). Not only would it be complicated to track which
+// goros are blocked on what, this use-case would introduce complicated
+// fairness/livelock issues that pathlock doesn't attempt to solve.
+type pathlock struct {
+	sync.Mutex
+	sync.Cond
+
+	// running contains the set of paths being written to by some in-progress
+	// PutFile operation (we use a set of sorted strings rather than a map so that
+	// we can quickly find prefix collisions)
+	running []string
+
+	// blocking is an in-progress PutFile operation that is blocking the next
+	// PutFile operation in the stream. Before exiting, each putFile goro checks
+	// if it's the blocking operation, and if so, signals to the blocked goro that
+	// it may proceed
+	blocking string
+
+	// This is used to limit the total concurrency of putFile request. It's
+	// included in pathlock as a convenience, as everywhere that putFile calls
+	// pathlock.start(), it would also call putFileLimiter.Acquire() (and likewise
+	// for pathlock.finish() and putFileLimiter.Release). This is unset for tests.
+	putFileLimiter limit.ConcurrencyLimiter
+}
+
+func newPathLock(putFileLimiter limit.ConcurrencyLimiter) *pathlock {
+	p := pathlock{
+		putFileLimiter: putFileLimiter,
+	}
+	p.Cond.L = &p.Mutex
+	return &p
+}
+
+func (p *pathlock) start(path string) (retErr error) {
+	// add leading slash, so '/' collides w/ everything.
+	// TODO(msteffen): would paths.go (in this package) or s/s/pkg/paths be
```

Comment:
For the storage v1 I think we use the clean from the hashtree package?

Reply:
:+1: I went through the process of trying to use that, and it actually ended up being a bit longer and more complicated than this (i.e. `"/"+strings.TrimPrefix(hashtree.Clean("/"+path), "/")`) so I decided to just remove the todo. Let me know if you think it's worth using hashtree anyway, but I figured shorter is better (and I removed the TODO)